File: 22-generate-parentheses/golang/solution.go
package golang

func generateParenthesis(n int) []string {
	var res []string
	generate(&res, n, 0, 0, "")
	return res
}

func generate(res *[]string, n, left, right int, s string) {
	if left == n && right == n {
		*res = append(*res, s)
	} else if left <= n && right <= n {
		if left < n {
			generate(res, n, left+1, right, s+"(")
		}
		if right < left {
			generate(res, n, left, right+1, s+")")
		}
	}
}

File: 22-generate-parentheses/python/solution.py
class Solution:
    def generateParenthesis(self, n: int) -> list[str]:
        def generate(left, right, s):
            if left == right == n:
                result.append(s)
            if left < n:
                generate(left + 1, right, s + '(')
            if right < left <= n:
                generate(left, right + 1, s + ')')

        result = []
        generate(0, 0, '')
        return result
